fc/code_generation.py

```diff
@@ -0,0 +1,284 @@
+
+    # Oxmeta namespace
+    oxmeta = 'https://chaste.comlab.ox.ac.uk/cellml/ns/oxford-metadata'
+
+    # Get unique names for all symbols
+    unames = get_unique_names(model)
+
+    # Symbol naming function
+    def symbol_name(symbol):
+        return 'var_' + unames[symbol]
+
+    # Derivative naming function
+    def derivative_name(deriv):
+        var = deriv.expr if isinstance(deriv, sympy.Derivative) else deriv
+        return 'd_dt_' + unames[var]
+
+    # Create expression printer
+    printer = WebLabPrinter(symbol_name, derivative_name)
+
+    # Create free variable name
+    free_name = symbol_name(model.get_free_variable_symbol())
+
+    # Create state information dicts
+    state_info = []
+    for i, state in enumerate(model.get_state_symbols()):
+        state_info.append({
+            'index': i,
+            'var_name': symbol_name(state),
+            'deriv_name': derivative_name(state),
+            'initial_value': model.get_initial_value(state),
+            'var_names': model.get_ontology_terms_by_symbol(state, oxmeta),
+        })
+
+    # Create parameter information dicts
+    parameter_info = []
+    for i, parameter in enumerate(parameters):
+        symbol = model.get_symbol_by_ontology_term(*parameter)
+        parameter_info.append({
+            'index': i,
+            'annotation': parameter,
+            'var_name': symbol_name(symbol),
+            'initial_value': model.get_value(symbol),
+        })
+
+    # Create map of parameter symbols to their indices
+    parameter_symbols = {}
+    for i, parameter in enumerate(parameters):
+        symbol = model.get_symbol_by_ontology_term(*parameter)
+        parameter_symbols[symbol] = i
+
+    # Create output information dicts
+    # Each output is associated either with a symbol or a parameter.
+    output_info = []
+    for i, output in enumerate(outputs):
+
+        # Allow special output value 'state_variable'
+        # TODO Replace this with a more generic implementation
+        if output == 'state_variable':
+            var_name = [{'index': x['index'], 'var_name': x['var_name']}
+                        for x in state_info]
+            parameter_index = None
+            length = len(state_info)
+        else:
+            symbol = model.get_symbol_by_ontology_term(*output)
+            var_name = symbol_name(symbol)
+            parameter_index = parameter_symbols.get(symbol, None)
+            length = None
+
+        output_info.append({
+            'index': i,
+            'annotation': output,
+            'var_name': var_name,
+            'parameter_index': parameter_index,
+            'length': length,
+        })
+
+    # Create RHS equation information dicts
+    rhs_equations = []
+    for eq in model.get_equations_for(model.get_derivative_symbols()):
+        # TODO: Parameters should never appear as the left-hand side of an
+        # equation (cellmlmanip should already have filtered these out).
```

Comment:
This is probably not true now... (but probably shouldn't fix in this PR?)

Reply:
Thanks. Made an issue for it https://github.com/ModellingWebLab/weblab-fc/issues/87